File: src/lib/inherit.ts
import { defaultAuth } from "./helpers";
import type { ApiRequest, AuthConfig, CollectionGroup, FolderConfig, KeyValue } from "@/types";

export function folderPathChain(folder?: string | null): string[] {
  const normalized = folder?.trim();
  if (!normalized) return [];
  const parts = normalized.split("/").filter(Boolean);
  return parts.map((_, index) => parts.slice(0, index + 1).join("/"));
}

export function folderConfigFor(collection: CollectionGroup | null, path: string): FolderConfig | undefined {
  return collection?.folderConfigs?.find((item) => item.path === path);
}

function isConcreteAuth(auth?: AuthConfig | null): auth is AuthConfig {
  return Boolean(auth && auth.authType && auth.authType !== "inherit" && auth.authType !== "none");
}

export type InheritSource = {
  label: string;
  auth: AuthConfig;
};

export function resolveInheritedAuth(
  requestAuth: AuthConfig,
  collection: CollectionGroup | null,
  folder?: string | null,
): InheritSource {
  if (requestAuth.authType !== "inherit") {
    return { label: "This request", auth: requestAuth };
  }

  const chain = folderPathChain(folder);
  for (let index = chain.length - 1; index >= 0; index -= 1) {
    const config = folderConfigFor(collection, chain[index]!);
    if (!config?.auth || config.auth.authType === "inherit") continue;
    if (config.auth.authType === "none") {
      return { label: `Folder ${chain[index]}`, auth: config.auth };
    }
    if (isConcreteAuth(config.auth)) {
      return { label: `Folder ${chain[index]}`, auth: config.auth };
    }
  }

  if (collection?.auth && collection.auth.authType !== "inherit") {
    return { label: `Collection ${collection.name}`, auth: collection.auth };
  }

  return { label: "No parent auth", auth: defaultAuth() };
}

function joinScripts(scripts: Array<string | undefined>): string {
  return scripts
    .map((script) => script?.trim() ?? "")
    .filter(Boolean)
    .join("\n\n");
}

export function resolveInheritedPreRequest(
  requestScript: string,
  collection: CollectionGroup | null,
  folder?: string | null,
): string {
  const folderScripts = folderPathChain(folder).map(
    (path) => folderConfigFor(collection, path)?.preRequestScript,
  );
  return joinScripts([collection?.preRequestScript, ...folderScripts, requestScript]);
}

export function resolveInheritedTests(
  requestTests: string,
  collection: CollectionGroup | null,
  folder?: string | null,
): string {
  const folderScripts = folderPathChain(folder)
    .slice()
    .reverse()
    .map((path) => folderConfigFor(collection, path)?.tests);
  return joinScripts([requestTests, ...folderScripts, collection?.tests]);
}

export function collectFolderVariables(
  collection: CollectionGroup | null,
  folder?: string | null,
): KeyValue[] {
  const rows: KeyValue[] = [];
  for (const path of folderPathChain(folder)) {
    const config = folderConfigFor(collection, path);
    if (config?.variables?.length) rows.push(...config.variables);
  }
  return rows;
}

export function applyRequestInheritance(
  request: ApiRequest,
  collection: CollectionGroup | null,
  folder?: string | null,
): ApiRequest {
  const auth = resolveInheritedAuth(request.auth, collection, folder).auth;
  return {
    ...request,
    auth,
    preRequestScript: resolveInheritedPreRequest(request.preRequestScript, collection, folder),
    tests: resolveInheritedTests(request.tests, collection, folder),
  };
}

export function upsertFolderConfig(
  collection: CollectionGroup,
  path: string,
  patch: Partial<FolderConfig>,
): CollectionGroup {
  const folderConfigs = [...(collection.folderConfigs ?? [])];
  const index = folderConfigs.findIndex((item) => item.path === path);
  const next: FolderConfig = {
    path,
    ...(index >= 0 ? folderConfigs[index] : {}),
    ...patch,
  };
  if (index >= 0) folderConfigs[index] = next;
  else folderConfigs.push(next);
  return { ...collection, folderConfigs };
}

File: crates/pulse-core/src/inherit.rs
use crate::types::{ApiRequestDto, CollectionDto, FolderConfigDto, RequestAuthDto};

fn folder_path_chain(folder: Option<&str>) -> Vec<String> {
    let Some(normalized) = folder.map(str::trim).filter(|value| !value.is_empty()) else {
        return Vec::new();
    };
    let parts: Vec<&str> = normalized.split('/').filter(|part| !part.is_empty()).collect();
    parts
        .iter()
        .enumerate()
        .map(|(index, _)| parts[..=index].join("/"))
        .collect()
}

fn folder_config_for<'a>(collection: &'a CollectionDto, path: &str) -> Option<&'a FolderConfigDto> {
    collection.folder_configs.iter().find(|item| item.path == path)
}

fn join_scripts(scripts: impl IntoIterator<Item = Option<String>>) -> String {
    scripts
        .into_iter()
        .flatten()
        .map(|script| script.trim().to_string())
        .filter(|script| !script.is_empty())
        .collect::<Vec<_>>()
        .join("\n\n")
}

pub fn resolve_inherited_auth(
    request_auth: &RequestAuthDto,
    collection: Option<&CollectionDto>,
    folder: Option<&str>,
) -> RequestAuthDto {
    if request_auth.auth_type != "inherit" {
        return request_auth.clone();
    }
    let Some(collection) = collection else {
        return RequestAuthDto {
            auth_type: "none".into(),
            ..RequestAuthDto::default()
        };
    };
    let chain = folder_path_chain(folder);
    for path in chain.into_iter().rev() {
        if let Some(config) = folder_config_for(collection, &path) {
            if let Some(auth) = &config.auth {
                if auth.auth_type == "inherit" {
                    continue;
                }
                return auth.clone();
            }
        }
    }
    if let Some(auth) = &collection.auth {
        if auth.auth_type != "inherit" {
            return auth.clone();
        }
    }
    RequestAuthDto {
        auth_type: "none".into(),
        ..RequestAuthDto::default()
    }
}

pub fn collect_folder_variables(
    collection: Option<&CollectionDto>,
    folder: Option<&str>,
) -> Vec<crate::types::EnvVariable> {
    let Some(collection) = collection else {
        return Vec::new();
    };
    let mut rows = Vec::new();
    for path in folder_path_chain(folder) {
        if let Some(config) = folder_config_for(collection, &path) {
            rows.extend(config.variables.clone());
        }
    }
    rows
}

pub fn apply_request_inheritance(
    mut request: ApiRequestDto,
    collection: Option<&CollectionDto>,
    folder: Option<&str>,
) -> ApiRequestDto {
    request.auth = resolve_inherited_auth(&request.auth, collection, folder);
    let folder_pre = folder_path_chain(folder)
        .into_iter()
        .map(|path| {
            collection
                .and_then(|col| folder_config_for(col, &path))
                .and_then(|config| config.pre_request_script.clone())
        });
    request.pre_request_script = join_scripts(
        std::iter::once(collection.and_then(|col| col.pre_request_script.clone()))
            .chain(folder_pre)
            .chain(std::iter::once(Some(request.pre_request_script))),
    );
    let folder_tests = folder_path_chain(folder)
        .into_iter()
        .rev()
        .map(|path| {
            collection
                .and_then(|col| folder_config_for(col, &path))
                .and_then(|config| config.tests.clone())
        });
    request.tests = join_scripts(
        std::iter::once(Some(request.tests))
            .chain(folder_tests)
            .chain(std::iter::once(collection.and_then(|col| col.tests.clone()))),
    );
    request
}
